packages/models-tree/src/models-tree/ModelsTreeDefinition.ts

```diff
@@ -740,40 +774,72 @@ async function createInstanceKeyPathsFromTargetItems({
-      } else if (await imodelAccess.classDerivesFrom(key.className, "BisCore.SpatialCategory")) {
-        ids.categories.push(key.id);
-      } else {
-        ids.elements.push(key.id);
-      }
-    }),
-  );
-  const elementBlocks: Array<Array<Id64String | ElementsGroupInfo>> = [];
-  const elementsLength = ids.elements.length;
-  const blockSize = Math.ceil(elementsLength / Math.ceil(elementsLength / 5000));
-  for (let i = 0; i < ids.elements.length; i += blockSize) {
-    elementBlocks.push(ids.elements.slice(i, i + blockSize));
-  }
-
-  return collect(
-    merge(
-      from(ids.subjects).pipe(mergeMap((id) => createSubjectInstanceKeysPath(id, idsCache))),
-      from(ids.models).pipe(mergeMap((id) => createModelInstanceKeyPaths(id, idsCache))),
-      from(ids.categories).pipe(mergeMap((id) => createCategoryInstanceKeyPaths(id, idsCache))),
-      from(elementBlocks).pipe(mergeMap((block) => createGeometricElementInstanceKeyPaths(imodelAccess, idsCache, hierarchyConfig, block))),
+  return lastValueFrom(
+    from(targetItems).pipe(
+      releaseMainThreadOnItemsCount(2000),
+      mergeMap(async (key): Promise<{ key: string; type: number } | { key: ElementsGroupInfo; type: 0 }> => {
+        if ("parent" in key) {
+          return { key, type: 0 };
+        }
+
+        const subjectCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.Subject");
+        const isSubject = subjectCheck instanceof Promise ? await subjectCheck : subjectCheck;
+        if (isSubject) {
+          return { key: key.id, type: 1 };
+        }
+
+        const modelCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.Model");
+        const isModel = modelCheck instanceof Promise ? await modelCheck : modelCheck;
+        if (isModel) {
+          return { key: key.id, type: 2 };
+        }
+
+        const spatialCategoryCheck = imodelAccess.classDerivesFrom(key.className, "BisCore.SpatialCategory");
+        const isSpatialCategoryCheck = spatialCategoryCheck instanceof Promise ? await spatialCategoryCheck : spatialCategoryCheck;
+        if (isSpatialCategoryCheck) {
+          return { key: key.id, type: 3 };
+        }
+
+        return { key: key.id, type: 0 };
+      }, 2),
+      toArray(),
+      switchMap(async (keysParsed) => {
+        const ids = {
+          models: new Array<Id64String>(),
+          categories: new Array<Id64String>(),
+          subjects: new Array<Id64String>(),
+          elements: new Array<Id64String | ElementsGroupInfo>(),
+        };
+        for (const keyParsed of keysParsed) {
+          if (keyParsed.type === 1) {
+            ids.subjects.push(keyParsed.key);
+            continue;
+          }
+          if (keyParsed.type === 2) {
+            ids.models.push(keyParsed.key);
+            continue;
+          }
+          if (keyParsed.type === 3) {
+            ids.categories.push(keyParsed.key);
+            continue;
+          }
+          ids.elements.push(keyParsed.key);
+        }
```

Comment:
why not have a `reduce` that creates the `ids` structure in the first place, instead of `toArray` to create an array and then convert it to `ids`?

Reply:
Changed